### PR TITLE
Remove checks for nullptr from BlockAssembler::CreateNewBlock

### PR DESCRIPTION
**Pull request by cculianu**
Motivation/rationale: Minor nit.

Background:

See: https://en.cppreference.com/w/cpp/memory/new/operator_new

`operator new()` does not normally ever return `nullptr` unless one
explicitly uses the `std::nothrow` version or unless one redefines it.  Neither
is the case in this entire codebase, hence the check for `nullptr` for the
return value from `BlockAssembler::CreateNewBlock` was entirely
superfluous and all branches that test it would always evaluate to
`false`. So, it can be safely removed in the interests of code quality.

tl;dr: `operator new()` either succeeds or throws `std::bad_alloc`, it can
never return `nullptr`.

### CODE REVIEW

**Comment by MarcoFalke** [comment]
If it is impossible to return a nullptr, then why not return a plain `CBlockTemplate`?

**Comment by cculianu** [comment]
@MarcoFalke Good question.  I'm not sure what the motivation is behind the `BlockAssembler` class and its design.  Outside of tests, it's only ever used in `getblocktemplate` and `generateBlocks` in `rpc/mining.cpp`.  It is always created as an ephemeral object on the stack and then after it is done creating a block template, it is immediately destroyed after the stack frame ends. 

Why it allocates a block template as a pointer and then returns it via a `unique_ptr` is a good question.  I guess because a block template is potentially very "heavy" and the people designing the subsystem didn't want to mess with `std::move` and/or rvalue reference `&&` semantics, and `unique_ptr` is just easier to think about for some people?  Or maybe this evolved from some older code that did it with raw pointers?  I don't know...  If I had done it perhaps I would have just used a `CBlockTemplate` and ensured it had proper `std::move` semantics and returned it directly by value, leveraging RVO and `&&` move semantics.

**Comment by MarcoFalke** [comment]
What I meant is that simply removing nullptr checks is scary and will also fail to compile on some compilers/sanitizers/static analysers with warnings and errors enabled. Also, if the template ever only exists on the stack, then I don't see why it should live in the heap.

I am not sure if there is a way in C++ to ensure an object can be efficiently moved, but a pre-C++11 alternative would be:

```cpp
CBlockTemplate on_the_stack;
BlockAssembler().CreateNewBlock(coinbase_script, on_the_stack);

**Comment by cculianu** [comment]
Yeah that was the pre-C++11 canonical way to do it.  Also just returning the value was pretty much always guaranteed to not produce any copies and be equivalent to the "pass down a non-const ref" way..  (RVO/copy elision basically does that for you internally anyway).  

The example you show is the cheapest way (as is a proper RVO return by value), for sure.  Move semantics are "almost free" too -- you just have to pay the extra cost of swapping pointers, etc (CBlockTemplate's heaviest members are the vector of tx refs and other stuff and so you pay the cost of that pointer swap...).

> What I meant is that simply removing nullptr checks is scary and will also fail to compile

:) It is only scary if you are thinking about it as just a fancy `malloc`. In C, definitely, not checking `NULL` is scary.  

However, in C++, `new` can never return `nullptr` (at least not the  `new` used in this entire codebase).  It's guaranteed to throw or succeed.  Those are its only two postconditions.  In light of that, there are no sanitizers or compilers that will error out on it or  complain -- since the standard in C++  specifies exactly what new does...
 
If you find this PR useless or scary, feel free to close it.  I figured rather than complain I could help out and fix the little nits here and there.

No worries either way.

**Comment by MarcoFalke** [comment]
> new can never return nullptr

Oh, I meant that compilers can't know that `CreateNewBlock` never returns a nullptr.

```cpp
const auto b = CreateNewBlock();
b->Get();  // <-- Warning/Error here
```

> If you find this PR useless

We do appreciate refactoring changes that improve the code base. My feedback was about the current version of the pull request, which I believe can not be merged as-is.

**Comment by cculianu** [comment]
Do you know of any static analyzer that would complain about this code?  The ones I use (PVS studio) do not necessarily expect you to check `unique_ptr` return value since that is a very common and idiomatic pattern -- call a factory method returning `unique_ptr`.  Since `new` throws, it's considered by some to be an anti-pattern to check it --  unless the contract with the calling code specifies "will return a nullptr on error", which is not the case here.  

I ran PVS studio on this codebase without my change -- and the check itself generated an issue (basically because nullptr can never be returned from new).

I ran it with this change and it did not complain.  I can paste the log...

**Comment by cculianu** [comment]
Just to reiterate -- code like this:

```c++
auto ptr = std::make_unique<MyClass>(arg1, arg2);
if (!ptr)
  // handle error here
```

Is unidiomatic and kind of silly... since it can never be false.  For that reason no static analyzer that I know of expects you to check `unique_ptr` return value either.

**Comment by MarcoFalke** [comment]
For our codebase, unique_ptr means that the pointer can be null. E.g.

* https://github.com/bitcoin/bitcoin/blob/195822f1e05e2f36002c906667d4c639663f23b5/src/node/context.h#L43
* https://github.com/bitcoin/bitcoin/blob/195822f1e05e2f36002c906667d4c639663f23b5/src/net.h#L835
* ...

So I think it would be confusing to have one exception where a pointer type is used but at the same time it can not be null.

**Comment by cculianu** [comment]
Right, but in this case the contract is that it cannot be null.

I get it -- this is a minor nit and not worth worrying about -- no worries.  I still maintain that it would be a code quality improvement.  It's a bit silly to see people checking `new`'s return value as if it can ever return `nullptr`.  :)

But I get it.  You don't want to worry about this.  Feel free to close this.  Like I said -- no harm done, no offense taken, and no worries.

**Comment by laanwj** [comment]
I agree with @MarcoFalke. Please make it either impossible to return a nullptr (by returning a while object) or keep the checks somehow. Simply removing them is scary.
Even though the check might, with some reasoning, be unnecessary right now, for maintainability and robustness some defense in depth doesn't hurt so NACK on this change as it is.

**Comment by cculianu** [comment]
> Please make it either impossible to return a nullptr (

My point is -- *It's already impossible to return a nullptr*.  Which is why the check is dumb.  It's the equivalent of doing:

```
   int a = 2;
   if (a != 2) {
      // .... code here to throw
   }
```

The above is an exaggeration, of course -- but if you stop to think about it -- it really is that silly.

Anyway, I'm closing this.  I feel like I'm  fighting against the wind here.  Good luck.
